### DISCUSSION TITLE
Micropython automatically converting stdin carriage returns to newlines?

### DISCUSSION THREAD

**Discussion by sandyscott** [micropython/micropython#9758] [Core Development] [answered]
I was looking to modify some terminal code (specifically, [aiorepl](https://github.com/micropython/micropython-lib/tree/master/micropython/aiorepl))to make it work nicely with windows clients and encountered this issue.

When I press enter in a windows terminal, MicroPython (v1.19.1 on a Raspberry Pi Pico) appears to recieve 2 linefeed characters (0x0A + 0x0A) instead of a carriage return then a linefeed (0x0D + 0x0A).

Code:

```
import sys

while(True):
    #input = sys.stdin
    b = sys.stdin.read(1)
    sys.stdout.write("(" + hex(ord(b)) + ")")
```

Run the code, connect to port using a [modified PuTTY](https://github.com/gniemirowski/putty-crlf) configured to send a CRLF, hit enter and you get:

![image](https://user-images.githubusercontent.com/2547618/198259026-5da3ccb2-b3d7-4e76-b21e-4135ac9bce14.png)

I get similar results when using the Thonny Shell or the REPL shell included in the Micropython plugin for PyCharm.

Am I doing something wrong here or is this a quirk (or bug) with MicroPython?

**Comment by dpgeorge** [comment, marked as answer]
Yes, MicroPython translates `\r` to `\n` (and I think this even follows how CPython works).

You can try using `sys.stdin.buffer.read(1)`, that will not do any translation.

**Reply by sandyscott** [reply]
OK, thank you. I couldn't find much in the official documentation until I found [PEP 278 – Universal Newline Support](https://peps.python.org/pep-0278/). I have submitted a [PR](https://github.com/micropython/micropython-lib/pull/559) for `aiorepl` which uses `.buffer.read()`, but having seen that PEP I'll see I can simplify it and stick with the standard behaviour.

**Reply by sandyscott** [reply]
New, much simpler PR: https://github.com/micropython/micropython-lib/pull/560
